parsl/providers/azure/azure.py

```diff
@@ -0,0 +1,478 @@
+    your azure account),
+    `AZURE_CLIENT_SECRET` (the secret key for your azure account), the
+    `AZURE_TENANT_ID` (the session key for your azure account), and
+    AZURE_SUBSCRIPTION_ID.
+
+    Parameters
+    ----------
+    vm_reference : dict
+        Dictionary describing the parameters of the Azure VM.
+
+        Required structure:
+        {
+          'publisher': VM OS publisher
+          'offer': VM OS offer
+          'sku': VM OS SKU
+          'version': VM OS version
+          'vm_size': VM Size, analogous to instance type for AWS
+          'disk_size_gb': (int) size of VM disk in gb
+          "admin_username": (str) admin username of VM instances,
+          "password": (str) admin password for VM instances
+        }
+    worker_init : str
+        String to append to the Userdata script executed in the cloudinit phase of
+        instance initialization.
+    key_file : str
+        Path to json file that contains 'Azure keys'
+    init_blocks : int
+        Number of blocks to provision at the start of the run. Default is 1.
+    min_blocks : int
+        Minimum number of blocks to maintain. Default is 0.
+    max_blocks : int
+        Maximum number of blocks to maintain. Default is 10.
+    region : str
+        Azure region to launch machines. Default is 'westus'.
+    key_name : str
+        Name of the Azure private key (.pem file) that is usually generated on the console
+        to allow SSH access to the Azure instances. This is mostly used for debugging.
+    launcher : Launcher
+        Launcher for this provider. Possible launchers include
+        :class:`~parsl.launchers.SingleNodeLauncher` (the default),
+        :class:`~parsl.launchers.SrunLauncher`, or
+        :class:`~parsl.launchers.AprunLauncher`
+    linger : Bool
+        When set to True, the workers will not `halt`. The user is responsible for shutting
+        down the nodes.
+    """
+
+    def __init__(self,
+                 init_blocks=1,
+                 min_blocks=0,
+                 max_blocks=10,
+                 parallelism=1,
+                 worker_init='',
+                 vm_reference=None,
+                 location='westus',
+                 group_name='parsl.auto',
+                 key_name=None,
+                 key_file=None,
+                 vnet_name="parsl.auto",
+                 linger=False,
+                 launcher=SingleNodeLauncher()):
+        if not _api_enabled:
+            raise OptionalModuleMissing(
+                ['azure', 'msrestazure'], "Azure Provider requires the azure module.")
+
+        self._label = 'azure'
+        self.init_blocks = init_blocks
+        self.min_blocks = min_blocks
+        self.max_blocks = max_blocks
+        self.max_nodes = max_blocks
+        self.parallelism = parallelism
+        self.nodes_per_block = 1
+
+        self.worker_init = worker_init
+        self.vm_reference = vm_reference
+        self.vm_disk_size = self.vm_reference["disk_size_gb"]
```

Comment:
Does this `self.vm_disk_size` need to be set? I can't see anywhere that it is used